pkg/otelcollector/config/provider.go

```diff
@@ -103,26 +89,54 @@ func (p *Provider) setConfig(config *Config) {
 // The hook should treat the given config as temporary.
 // The hook will also be executed immediately, to ensure that current config
 // was passed through all the added hooks.
+//
+// WARNING: This is supposed to be called only during initialization.
 func (p *Provider) AddMutatingHook(hook func(*Config)) {
-	p.lock.Lock()
-	defer p.lock.Unlock()
-
 	if p.config == nil {
 		log.Warn().Msg("OtelConfigProvider.AddHook: already shut down")
 		return
 	}
 
 	p.hooks = append(p.hooks, hook)
 
-	// Now the config provided to collector is outdated, run the newly-added hook.
-	hook(p.config)
-	p.setConfig(p.config)
+	p.UpdateConfig(p.config)
+}
+
+// GetConfig returns the current config.
+func (p *Provider) GetConfig() *Config {
+	p.configLock.RLock()
+	defer p.configLock.RUnlock()
+	return p.config
+}
+
+// SetConfig sets the new config, replacing the old one.
+func (p *Provider) SetConfig(config *Config) {
```

Comment:
Why all the "internal" functions are public? This just begs for breaking invariants